pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy

```diff
@@ -359,7 +361,17 @@ class ModelValidatorImpl implements ModelValidator {
     private boolean validateStep(ModelASTStep step, DescribableModel<? extends Describable> model, Descriptor desc) {
 
         if (step instanceof AbstractModelASTCodeBlock) {
-            // No validation needed for code blocks like expression and script
+            // Verify that the code block can be parsed - we'll still get garbage for errors around class imports, etc,
+            // but you can't do that from the editor anyway.
+            String codeBlock = step.codeBlockAsString()
+            CompilationUnit cu = new CompilationUnit()
+            cu.addSource(step.name, codeBlock)
+            try {
+                cu.compile(Phases.PARSING)
+            } catch (CompilationFailedException e) {
+                errorCollector.error(step, Messages.ModelValidatorImpl_CompilationErrorInCodeBlock(step.name))
```

Comment:
Can we get a line number or something from the unused exception here?

Reply:
You don't actually want the line number - that's from the converted-to-Groovy version. You *are* getting the JSON location, though.

Reply:
(i.e., same as any other validation error on a JSON element)

Reply:
What about any other information? I'll show an error to the user but it should have some context about what went wrong so the user can correct it

Reply:
See my note in the description - there's no way for me to get a more granular error than `startup failed` without digging into the full error string a la 
```
startup failed:
script: 1: unexpected char: ''' @ line 1, column 63.
    "Sleeping 1S"; sleep 1;' done'
                                 ^
```
If you want that full message, I'll happily give it to you, but I assumed you wouldn't want it due to the size.

Reply:
Yeah, would rather have the error than not.

Reply:
Wait, I may have found a way...

Reply:
Ok, the error message is now gonna be like:
```
Groovy compilation error(s) in script. Error(s): "unexpected char: '''"
```
In addition to the JSON location, of course.